### DISCUSSION TITLE
Using this as a Library

### DISCUSSION THREAD

**Discussion by edrose** [ExpressLRS/ExpressLRS#1866] [Q&A] [answered]
Hi all,

Thanks for the great project. I'm just getting into building my first DIY drone and it's amazing to see projects like this within the space.

I'm currently designing my own custom flight computer as a personal project, which will be based off a high performance STM32. I was looking at incorporating an ELRS receiver into the flight computer to keep the whole think neat and tidy. I just had a few questions for someone who is familiar with the codebase:

1. Is there any way currently to build the code as a library, so it can be incorporated into the firmware of the flight controller MCU without needing a second MCU?
2. If not, how modular is the code and how easy would it be to modify?
3. If I were to put some effort into achieving this, is it a desirable feature for the project that I could contribute to?

I realise that my use case is likely very niche, but I thought I'd ask the question before I finalise my PCB design.

All the best,
Ed

**Comment by pkendall64** [comment, marked as answer]
I'll answer your questions first, then give a suggestion.

1. ExpressLRS is not build as a library and can't really be built as it assumes that it is going to be in control of timers and SPI devices as a minimum to get things going.
2. The code is broken into smaller modules, but the main bulk of the radio side of the code is in the main file rx_main.cpp.
3. Whilst making the code more modular/portable is a good goal, and one we have been working towards, the goal is not to make the code run inside another project but rather one of maintainability.

The devs currently prefer that new flight controllers with inbuilt ELRS use a serial connected ESP8285 based design. Some of the reasons for this are:

1. User upgradable so users can test PRs, branches, point-releases.
2. Upgrades don't have to wait for a new release of FC software.
3. Wifi upgradable and configuration (you don't get that on STM32).
4. Integrating ELRS into FC software really means that the FC software is ELRS compatible rather than running ELRS, see Betaflight as an example. That have to be completely done from scratch to fit in with the Betaflight codebase.

**Comment by edrose** [comment]
Thank you for that reply, that answers all of my questions. I'll take a look at the code, consider your suggestion, and work out what the best option is in my case.

I hope you have a good week.
Ed
